File: ch02/python/2.5.pow.py
def pow_iter(base, exp):

    result = 1

    for _ in range(exp):
        result *= base

    return result

print(pow_iter(3, 6))
print(3 ** 6)

def pow_recur(base, exp):

    if exp == 1:
        return base

    if exp % 2 == 0:
        result = pow_recur(base, exp / 2)
        return result * result

    result = pow_recur(base, (exp - 1) / 2)
    return base * result * result

print(pow_recur(3, 6))
print(pow_recur(17, 6))

File: ch02/nodejs/2.5.pow.js
const powIter = (base, exp) => {
    let result = 1;

    while (exp > 0) {
        result = result * base
        exp--;
    }

    return result;
}

console.log(powIter(3, 6))
console.log(powIter(10, 3))
console.log(powIter(17, 10))

const powRecur = (base, exp) => {
    if (exp === 1) return base
    let result;

    if (exp % 2 === 0) {
        result = powRecur(base, exp / 2);
        return result * result
    }

    result = powRecur(base, (exp - 1) / 2);
    return base * result * result
}

console.log(powRecur(3, 6))
console.log(powRecur(17, 10))
console.log(powRecur(3, 500))